lib/Geo/Coding.cpp

```diff
@@ -0,0 +1,317 @@
+  TRI_ASSERT(S2::IsUnitLength(point));
+  S2LatLng const latLng{point};
+  TRI_ASSERT(latLng.is_valid());
+  auto const lat = ((latLng.lat().radians() + M_PI_2) * kMax) / M_PI;
+  TRI_ASSERT(lat <= kMax) << lat;
+  encoder.put32(static_cast<uint32>(lat));
+  auto const lng = ((latLng.lng().radians() + M_PI) * kMax) / (2.0 * M_PI);
+  TRI_ASSERT(lng <= kMax) << lng;
+  encoder.put32(static_cast<uint32>(lng));
+}
+
+void decodePointFromLatLngInt(Decoder& decoder, S2Point& point) {
+  TRI_ASSERT(decoder.avail() >= 2 * sizeof(uint32));
+  auto const lat = static_cast<double>(decoder.get32());
+  auto const lng = static_cast<double>(decoder.get32());
+  auto const latLng = S2LatLng::FromRadians(
+      ((lat * M_PI) / kMax) - M_PI_2, ((lng * (2.0 * M_PI)) / kMax) - M_PI);
+  TRI_ASSERT(latLng.is_valid());
+  point = latLng.ToPoint();
+  TRI_ASSERT(S2::IsUnitLength(point));
+}
+
+void encodePointToLatLng(Encoder& encoder, S2Point const& point) {
+  TRI_ASSERT(encoder.avail() >= 2 * sizeof(double));
+  TRI_ASSERT(S2::IsUnitLength(point));
+  S2LatLng const latLng{point};
+  TRI_ASSERT(latLng.is_valid());
+  encoder.putdouble(latLng.lat().radians());
+  encoder.putdouble(latLng.lng().radians());
+}
+
+void decodePointFromLatLng(Decoder& decoder, S2Point& point) {
+  TRI_ASSERT(decoder.avail() >= 2 * sizeof(double));
+  auto const lat = decoder.getdouble();
+  auto const lng = decoder.getdouble();
+  auto const latLng = S2LatLng::FromRadians(lat, lng);
+  TRI_ASSERT(latLng.is_valid());
+  point = latLng.ToPoint();
+  TRI_ASSERT(S2::IsUnitLength(point));
+}
+
+void encodePointToPoint(Encoder& encoder, S2Point const& point) {
+  static_assert(sizeof(S2Point) == 3 * sizeof(double));
+  TRI_ASSERT(encoder.avail() >= sizeof(S2Point));
+  TRI_ASSERT(S2::IsUnitLength(point));
+  encoder.putn(&point, sizeof(S2Point));
+}
+
+void decodePointFromPoint(Decoder& decoder, S2Point& point) {
+  TRI_ASSERT(decoder.avail() >= 3 * sizeof(double));
+  point[0] = decoder.getdouble();
+  point[1] = decoder.getdouble();
+  point[2] = decoder.getdouble();
+  TRI_ASSERT(S2::IsUnitLength(point));
+}
+
+using namespace coding;
+
+void encodeVertices(Encoder& encoder, std::span<S2Point const> vertices,
+                    Options::Point options) {
+  switch (options) {
+    case Options::Point::S2LatLngInt: {
+      encoder.Ensure(vertices.size() * 2 * sizeof(uint32));
+      for (auto const& vertex : vertices) {
+        encodePointToLatLngInt(encoder, vertex);
+      }
+    } break;
+    case Options::Point::S2LatLng: {
+      encoder.Ensure(vertices.size() * 2 * sizeof(double));
+      for (auto const& vertex : vertices) {
+        encodePointToLatLng(encoder, vertex);
+      }
+    } break;
+    case Options::Point::S2Point: {
+      encoder.Ensure(vertices.size() * sizeof(S2Point));
+      for (auto const& vertex : vertices) {
+        encodePointToPoint(encoder, vertex);
+      }
+    } break;
+    default: {
+      TRI_ASSERT(false) << "Not implemented yet.";
```

Comment:
NIT add there what option exactly is not implemented

Reply:
Done